core/src/test/java/nu/yona/server/util/LockPoolTest.java

```diff
@@ -0,0 +1,165 @@
+package nu.yona.server.util;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.Test;
+
+import nu.yona.server.exceptions.YonaException;
+
+public class LockPoolTest
+{
+	private final AtomicInteger freeLockID = new AtomicInteger(0);
+
+	@Test
+	public void testLockingNonReentrant()
+	{
+		final int numThreads = 10;
+		final int iterations = 10000;
+		assertThat(testConcurrently(numThreads, iterations, this::attemptNonReentrantAccess), equalTo(0));
+	}
+
+	@Test
+	public void testLockingReentrant()
+	{
+		final int numThreads = 10;
+		final int iterations = 10000;
+		assertThat(testConcurrently(numThreads, iterations, this::attemptReentrantAccess), equalTo(0));
+	}
+
+	@Test
+	public void testConcurrencyDifferentKeys()
+	{
+		final int numThreads = 10;
+		final int iterations = 5;
+		assertThat(testConcurrently(numThreads, iterations, this::attempImmediateAccessRandomID), equalTo(0));
+	}
+
+	private int testConcurrently(int numThreads, int iterations, AccessAttempt accessAttempt)
+	{
+		LockPool<Integer> lockPool = new LockPool<>();
+		ExecutorService threadPool = Executors.newFixedThreadPool(numThreads);
+		CountDownLatch startSignal = new CountDownLatch(1);
+		CountDownLatch doneSignal = new CountDownLatch(numThreads);
+		AtomicBoolean concurrencyIndicator = new AtomicBoolean(false);
+		AtomicInteger failureCounter = new AtomicInteger(0);
```

Comment:
Because the `failureCounter` is incremented at multiple places, it would be easier for test fixing to use a failure log which you push a message to. It doesn't matter much for the completeness of the test though.